server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,314 @@
+            avg_delta = abs(avg - team_avg)
+            self._logger.debug("delta with %s is %s (avg is %s)", [item], avg_delta, team_avg)
+            if avg_delta < old_avg_delta:
+                candidate = item
+                old_avg_delta = avg_delta
+        self._logger.debug("used %s as filler", [candidate])
+        return candidate
+
+    def assign_game_quality(self, match: Match) -> GameCandidate:
+        newbie_bonus = 0
+        time_bonus = 0
+        ratings = []
+        for team in match:
+            for search in team.get_original_searches():
+                ratings.append(search.average_rating)
+                search_time_bonus = search.failed_matching_attempts * config.TIME_BONUS * len(search.players)
+                time_bonus += min(search_time_bonus, config.MAXIMUM_TIME_BONUS)
+                search_newbie_bonus = search.failed_matching_attempts * config.NEWBIE_BONUS * search.has_newbie()
+                newbie_bonus += min(search_newbie_bonus, config.MAXIMUM_NEWBIE_BONUS)
+
+        rating_disparity = abs(match[0].cumulated_rating - match[1].cumulated_rating)
+        fairness = max((config.MAXIMUM_RATING_IMBALANCE - rating_disparity) / config.MAXIMUM_RATING_IMBALANCE, 0)
+        deviation = statistics.pstdev(ratings)
+        uniformity = max((config.MAXIMUM_RATING_DEVIATION - deviation) / config.MAXIMUM_RATING_DEVIATION, 0)
+
+        quality = fairness * uniformity + newbie_bonus + time_bonus
+        self._logger.debug(
+            "bonuses: %s rating disparity: %s -> fairness: %f deviation: %f -> uniformity: %f -> game quality: %f",
+            newbie_bonus + time_bonus, rating_disparity, fairness, deviation, uniformity, quality)
+        return GameCandidate(match, quality)
+
+    def pick_noncolliding_games(self, games: List[GameCandidate]) -> List[Match]:
+        """
+        This greedily picks all matches with disjoint players, starting with the game with the highest quality.
+        This can miss more optimal solutions, but extensive testing showed that over many matchmaker
+        iterations there is no benefit to use a more sophisticated algorithm.
+        """
+        games = [game for game in games if game.quality >= config.MINIMUM_GAME_QUALITY]
+        self._logger.debug("%i games left after removal of games with quality < %s",
+                           len(games), config.MINIMUM_GAME_QUALITY)
+        games = SortedList(games, key=lambda game: game.quality)
+
+        matches = []
+        while games:
+            g = games.pop()
+            matches.append(g.match)
+            used_players = set(player for search in g.match for player in search.players)
+            self._logger.debug("used players: %s", [p.login for p in used_players])
+            games = [
+                game for game in games
+                if used_players.isdisjoint(
+                    player for search in game.match for player in search.players
+                )
```

Comment:
Maybe it would be worth refactoring this to a function or property so you could do something like:
```suggestion
                if used_players.isdisjoint(game.all_players)
```